### DISCUSSION TITLE
coerce fn() to fn() capturing

### DISCUSSION THREAD

**Discussion by TeamPuzel** [modular/modular#678] [Ideas]
When writing a List implementation this became a problem for functions like `map` `filter` or `fold`:
```nim
fn first(self, where: fn(T) -> Bool) -> Maybe[T]:
    for item in self:
        if where(item): return Maybe[T](item)
    return Maybe[T]()
    
fn first(self, where: fn(T) capturing -> Bool) -> Maybe[T]:
    for item in self:
        if where(item): return Maybe[T](item)
    return Maybe[T]()
```
I need to copy and paste every method for every combination of function parameters which is hard to maintain. I would expect all functions to work as parameters when `capturing` is present which turned out not to be true.

**Comment by lattner** [comment]
@laszlokindrat FYI we need to build support for subtyping of function types

**Reply by Moosems** [reply]
I would love to see the `type()` function come to python along with the Pipe operator (`|`) such that overloading doesn't create a bunch of boilerplate 😄.

**Reply by Moosems** [reply]
Much better to overload in scenarios where the code will be completely different and allow pipes and type() for areas where the code will remain relatively similar and one might just have to type cast.
